Cue/app/reducers/library.js

```diff
@@ -162,46 +162,129 @@ function library(state: State = initialState, action: Action): State {
+          mergedCards.splice(change.position, 0, card)
+        } else {
+          // The change is not a move; just edit the card in-place
+          mergedCards[index] = {...mergedCards[index], ...change, action: undefined}
+        }
+      } else {
+        console.warn('Tried to edit non-existant card ' + change.uuid + '?')
+      }
     }
   })
+
+  // Fix the position property of the cards if necessary
+  if (positionUpdateNeeded) {
+    mergedCards.forEach((card: Card, index: number) => {
+      card.position = index
+    })
+  }
+
   return mergedCards
 }
 
 //apply card changes to a localChange.cards
 function _applyCardChangesLocalChanges(cards, changes) {
   console.debug('_applyCardChangesLocalChanges', cards, changes)
-  if (! cards && !changes) return []
+  if (!cards && !changes) return []
   if (!cards) return changes
   if (!changes) return cards
   let mergedCards = cards.slice();
   changes.forEach(change => {
+    // Locate the existing localChanges record
     let index = mergedCards.findIndex(card => card.uuid == change.uuid)
+
     if (change.action === 'add') {
       mergedCards.push(change)
-    } else if (change.action === 'delete' && mergedCards[index]) {
-      if (mergedCards[index].action === 'add')
-        mergedCards.splice(index,1)
-      else
-        mergedCards[index] = {uuid: change.uuid, action: 'delete'}
+
+    } else if (change.action === 'delete') {
+      if (index !== -1) {
+        if (mergedCards[index].action === 'add') {
+          mergedCards.splice(index,1)
+        } else {
+          mergedCards[index] = {uuid: change.uuid, action: 'delete'}
+        }
+      } else {
+        mergedCards.push(change)
+      }
+
     } else if (change.action === 'edit') {
-      //TODO: make sure position order is what server expects
-      if (mergedCards[index])
-        mergedCards[index] = {...mergedCards[index], ...change, action: mergedCards[index].action}
-      else
+      // For card moves, since our local changes and the server's changes are
+      // applied in order, we only need to account for the case where a card
+      // is edited or moved multiple times, which causes other cards to shift.
+      if (index !== -1) {
+        if (typeof change.position !== 'undefined' && change.position !== index) {
+          // Remove the card from the list of merged cards
+          let localChange = mergedCards.splice(index, 1)[0]
+
+          let oldPosition = index
+          let newPosition = change.position
+
+          console.debug('Processing move from ' + oldPosition + ' to ' + newPosition)
+          // Forward move from X to Y; subtract 1 from all pending changes
+          // with positions in [X, Y)
+          if (newPosition > oldPosition) {
+            mergedCards.forEach((card: Card) => {
+              if (typeof card.position !== 'undefined'
+                  && card.position >= oldPosition
+                  && card.position < newPosition) {
```

Comment:
I don't think so, since if an object is at index `Y`, then since the new change is pushed onto the back of the array, when we process the changes, the object at index `Y` will just be bumped out of the way, which is what we expect.